libtenzir/builtins/formats/syslog.cpp

```diff
@@ -619,19 +618,30 @@ auto parse_loop(generator<std::optional<std::string_view>> lines,
     finish_all();
     builder.template emplace<Builder>();
   };
+  const auto length = [&]() {
+    return std::visit(
+      [](const auto& b) {
+        return b.rows_.size();
+      },
+      builder);
+  };
   auto line_nr = size_t{0};
-  auto last_finish = time::clock::now();
+  auto last_line_received = time::clock::now();
+  co_yield {};
   for (auto&& line : lines) {
-    for (auto&& slice : msb.yield_ready_as_table_slice()) {
-      co_yield std::move(slice);
-    }
     // We need our own timeout logic here, because we dont directly write events
     // into the MSB. Only on a `finish_all` or `finish_all_but_last` call events
     // are actually written into the MSB.
     auto now = time::clock::now();
-    if (now - last_finish > timeout) {
+    if (now - last_line_received > opts.settings.timeout) {
+      finish_all();
+      // We call finalize here because we do the timeout handling ourselves.
+      // Otherwise we would double the timeout.
+      for (auto&& slice : msb.finalize_as_table_slice()) {
+        co_yield std::move(slice);
+      }
+    } else {
```

Comment:
Don't you also need a "last_finish" logic, because you do want to finish eventually even when new lines come in?

Maybe: If `now > last_finish + timeout`, check `now > last_line_received + timeout`. If that holds true, finish all. Otherwise finish all but the last.